### DISCUSSION TITLE
comptime: if type is array

### DISCUSSION THREAD

**Discussion by danieldaeschle** [vlang/v#11797] [Ideas]
I need a feature (for json2) to check if type is an array and then checking the elem type of that.
```v
fn test<T>() ?T {
	mut data := T{}
	$for field in T.fields {
		$if field.typ is array {
			$if field.elem_type is string {
				data.$field.name = ['']
			}
		}
	}
	return data
}

struct User {
	test []string
}

fn main() {
	x := decode<User>()?
	println(x)
}
```

same for maps

**Comment by ntrel** [comment]
> 		$if field.typ is array {

I don't think it should work like that, because in V typed arrays do not have type `array`. `X is Y` means a type test. Allowing a special case would mean we then would not be able to detect when the type is literally `array` (untyped elements, which I think is still allowed but maybe shouldn't be). 

>			$if field.elem_type is string {

`field` should not have an `elem_type` field because all fields do not have an element type.

BTW `field.typ` is currently an int, but it should be a type literal, `field.Type` that can be used as a type e.g. for casting. (See #5972).

We could have type introspection:
```v
$if field.Type.kind == .array {
    $if field.Type.ElementType is string { // error if field.Type doesn't have an element type
```

Or, if compile-time evaluation of string expressions/methods is implemented we could rely on testing the type name:
```v
$if field.Type.name.starts_with('[]') {
    $if field.Type.name[2..] == 'string' {
```
This would be slower and might encourage users to try to parse complex types like maps wrongly.

> data.$field.name = ['']

This should be `data.$(field.name)` - see the docs:
https://github.com/vlang/V/blob/master/doc/docs.md#compile-time-reflection

**Comment by ntrel** [comment]
For a map field:
```v
$if field.Type.kind == .map {
    $if field.Type.KeyType is string { // error if field is not a map
```
Note that `Type.kind` would be different from `v.table.TypeSymbol.kind`:
```v
$if field.Type.kind == .string { // error, `string` is not a member of TypeKind, use `is` instead
```
```v
$if field.Type.kind == .pointer {
    $if field.Type.ElementType is string { // &string
```
```v
$if field.Type.kind == .optional {
    $if field.Type.ElementType is string { // ?string
    } $else $if field.Type.ElementType.kind == .map { // ?map[X]Y
```

**Comment by danieldaeschle** [comment]
> This should be data.$(field.name) - see the docs:
https://github.com/vlang/V/blob/master/doc/docs.md#compile-time-reflection

docs seem to be wrong. this was discussed with Alex and it's fine. It works this way.

**Comment by danieldaeschle** [comment]
i like your idea.. basically.
But `field.typ.kind == .array` is not a smartcast. So `field.typ.element_type` should work either always or never. But I get your pointer.

I also thought about having element_type on field which is 0 if the field has no element_type... but I also don't like that because of the reason you mentioned.

One thing I would appreciate is: We have methods on our Type type. What about using the predefined `is_array()`, `is_optional()` etc? But then we still have no way to access the `element_type`.

**Comment by ntrel** [comment]
> field.typ.kind == .array is not a smartcast. So field.typ.element_type should work either always or never

Yes it's not a smart cast, it's a property of a type. Types are different from struct instances, so we could have a magic `ElementType` property that is only valid to use if the type is an array, pointer or optional. The `KeyType` property would only work if the type is a map. Otherwise these would be compile-time errors. 

> What about using the predefined is_array(), is_optional() 

I thought of that too, but it keeps the language simpler to only have a `Type.kind` property and test its value at compile-time.

**Comment by danieldaeschle** [comment]
I'm just not a fried of compile time magic. I would love to have a feature which works easily without adjusting the compiler for every new field added.
